### PR TITLE
feat(ado): Sprint Poker Estimates

### PR DESCRIPTION
**Pull request by bobjac**
This pull request includes functionality for importing Azure DevOps work items into a sprint poker meeting and pushing estimates back to Azure Dev Ops.  It includes:

- Import user stories from Azure DevOps.  The backend code will look at the OAuth2 token provided by Azure Active Directory and query across all tenants and projects to find user stories the user is able to access.
- Those user stories can be imported
- More granular work items (tasks, bugs) can be found through search which is implemented in another PR by @danielkon96 
- Once the stories are imported, the can be estimated in the Poker meeting.  You can select the field you wish to push the estimates to (comment or base field for the work item type)

Screen Shot 2022-04-28 at 1 47 53 PM
Screen Shot 2022-04-28 at 1 48 37 PM
a552b.png)
Screen Shot 2022-04-28 at 1 49 28 PM
Screen Shot 2022-04-28 at 1 49 45 PM
Screen Shot 2022-04-28 at 1 50 12 PM
Screen Shot 2022-04-28 at 1 50 41 PM
Screen Shot 2022-04-28 at 1 51 05 PM

### CODE REVIEW

File: packages/server/utils/AzureDevOpsServerManager.ts
Comment (nickoferrall) [inline packages/server/utils/AzureDevOpsServerManager.ts:525]: -1 we're using markdown here but it looks like the comment is just a string in Azure:

comment-style

We try to keep PRs as small as possible so would you mind creating a separate issue to make these comments look more similar to our comments in the other integrations? For example, here's GitHub's:

Screenshot 2022-05-05 at 16 01 59

**Review by nickoferrall — CHANGES_REQUESTED** [review]
Looking good! One issue I noticed while testing is I'm unable to select the "effort" label on an Epic:
effort-label-bug

Also, it doesn't look like this is related to this PR, but I noticed that when you add an issue to the Estimate stage, it briefly just shows an id rather than a title:

opt-upd-err

It looks like the optimistic updater in [UpdatePokerScopeMutation](https://github.com/ParabolInc/parabol/blob/2c9c152df2ba74ae01f468f63f9168783146f84d/packages/client/mutations/UpdatePokerScopeMutation.ts#L220) needs some more info. It'd great if you or someone else from the team could create a separate issue/PR to fix this.

File: packages/client/components/AzureDevOpsFieldDimensionDropdown.tsx
Comment (bobjac) [reply packages/client/components/AzureDevOpsFieldDimensionDropdown.tsx:45]: Sounds good.  I'll change

File: packages/client/components/AzureDevOpsFieldDimensionDropdown.tsx
Comment (bobjac) [reply packages/client/components/AzureDevOpsFieldDimensionDropdown.tsx:89]: I will change.

File: packages/server/utils/AzureDevOpsServerManager.ts
Comment (wdfox) [reply packages/server/utils/AzureDevOpsServerManager.ts]: Fixed

File: packages/server/utils/AzureDevOpsServerManager.ts
Comment (wdfox) [reply packages/server/utils/AzureDevOpsServerManager.ts:166]: The Azure DevOps REST endpoint that we are calling returns an object that looks something like the following, so we chose to match that format at the top level:

```
{
  "id": 299,
  "rev": 6,
  "fields": {
      .....
  },
  "_links": {
    .....
  },
  "url": "https://dev.azure.com/fabrikam/_apis/wit/workItems/299"
}
```

Comment (bobjac) [reply packages/server/utils/AzureDevOpsServerManager.ts:166]: The Azure DevOps REST API is a hypermedia API, so that is why the _links property is there.

File: packages/server/utils/AzureDevOpsServerManager.ts
Comment (bobjac) [reply packages/server/utils/AzureDevOpsServerManager.ts:524]: We created issue https://github.com/ParabolInc/parabol/issues/6509.

We will resolve this in another PR as requested.  SHouldn't take more than an hour or so.

File: packages/client/components/AzureDevOpsFieldMenu.tsx
Comment (bobjac) [reply packages/client/components/AzureDevOpsFieldMenu.tsx:117]: Fixed

File: packages/server/utils/AzureDevOpsServerManager.ts
Comment (bobjac) [reply packages/server/utils/AzureDevOpsServerManager.ts]: Removed

File: packages/server/graphql/types/AzureDevOpsDimensionField.ts
Comment (bobjac) [reply packages/server/graphql/types/AzureDevOpsDimensionField.ts:32]: Resolved

File: packages/client/components/PokerEstimateHeaderCard.tsx
Comment (bobjac) [reply packages/client/components/PokerEstimateHeaderCard.tsx:48]: changed

File: packages/client/hooks/useMakeStageSummaries.ts
Comment (bobjac) [reply packages/client/hooks/useMakeStageSummaries.ts:108]: changed

File: packages/client/shared/gqlIds/AzureDevOpsDimensionFieldId.ts
Comment (bobjac) [reply packages/client/shared/gqlIds/AzureDevOpsDimensionFieldId.ts:3]: Comments added

File: packages/server/dataloader/azureDevOpsLoaders.ts
Comment (bobjac) [reply packages/server/dataloader/azureDevOpsLoaders.ts:367]: Changed

**Comment by bobjac** [comment]
Looks like some merge conflicts were introduced after the query PR was accepted/merged.  I am resolving now.

**Comment by bobjac** [comment]
Merge conflicts should be resolved

**Review by nickoferrall — CHANGES_REQUESTED** [review]
A few issues I've noticed while playing around with the integrations. Feel free to create new issues for the first two:

1. When I returned to the meeting, I couldn't see any issues. My server logs told me that I needed to refresh the token. How can we fix this so that users won't need to refresh their tokens so frequently?
2. If I click on the hyperlink in the Estimate task card (#2 in screenshot below), it doesn't take me to the Azure Devops project:
Screenshot 2022-05-06 at 16 14 58

3. After pulling the latest changes after you fixed the merge conflicts, I get a server error when pushing a score to any Azure Estimate item. 

<details>

```
16:19:26 1|GraphQL Executor  | Trace: {
16:19:26 1|GraphQL Executor  |   error: '{"data":null,"errors":[{"message":"column \\"azureDevOpsFieldName\\" of relation \\"TaskEstimate\\" does not exist","locations":[{"line":4,"column":3}],"path":["setTaskEstimate"]}]}'
16:19:26 1|GraphQL Executor  | }
16:19:26 1|GraphQL Executor  |     at executeGraphQL (/Users/nickoferrall/Parabol/dev/gqlExecutor.js:524:13)
16:19:26 1|GraphQL Executor  |     at processTicksAndRejections (node:internal/process/task_queues:96:5)
16:19:26 1|GraphQL Executor  |     at async onMessage (/Users/nickoferrall/Parabol/dev/gqlExecutor.js:98113:22)
16:19:26 2|Socket Server     | SEND TO SENTRY [Error: column "azureDevOpsFieldName" of relation "TaskEstimate" does not exist] undefined
```

</details>

4. This is unrelated to this PR, but in the Scope Search we have a placeholder `Search issues on Azure DevOps`. I just wanted to check whether everything would be referred to as an issue in Azure? I know we return issues but also tasks, epics, and user stories.

**Comment by bobjac** [comment]
Hey @nickoferrall - For item #3 (the azureDevOpsFieldName of relation 'TaskEstimate'), this is happening because @wdfox fixed a spelling type to the database column.  @wdfox  can confirm here, but I believe you will need to re-setup your local database container to run the migration.

I planned on creating separate issues for the first 2.  For item #2, there is a hypermedia field in the Azure DevOps api that returns the html url. I was planning on expanding the graph schema and capturing the html url on the work item so it could be displayed as that link.  The standard url, that is showing, is the hypermedia url mean for consumption by REST clients.

For the authentication/refresh token.....AAD / AzureDevOps has an expiration on the refresh token where JIRA doesn't.  At some point (which can be controlled by policy up to a certain amount), the refresh token is going to expire and the user will be forced to re-authenticate.

**Comment by bobjac** [comment]
@nickoferrall  - for you comment on #4 (Search issues on Azure DevOps).  An issue in Azure DevOps is a work item type.  In GitHub, the base type is Issue and labels are applied (bug, enancement).  In Azure DevOps, the was type is work item and there are actual inherited types from that (User Story, Issue, bug) and those inherited types are different depending on the project type.  So, there are "Issues" but it would make more sense from an Azure DevOps perspective to say "Work Items".  

I kept issues there because I was thinking in terms of how Parabol sees the base type.  I wasn't sure if parabol users consider everything an Issue and when you import from an external system, that issue is really a pointer to the type in the external system. 

It is an easy change that I would be happy to make if you think it should say "Work Items" in the case of Azure Dev Ops.

File: packages/server/postgres/migrations/1650768593875_add-AzureDevOpsDimensionFieldMap.ts
Comment (nickoferrall) [inline packages/server/postgres/migrations/1650768593875_add-AzureDevOpsDimensionFieldMap.ts:8]: +1 could change to `"id" SERIAL UNIQUE` to ensure it's a unique id

**Review by nickoferrall — APPROVED** [review]
LGTM! For the problem where we can't select the "effort" label for Epics, is this the issue that will resolve that: https://github.com/ParabolInc/parabol/issues/6512 ? 

And I just wanted to check if you've created issues for the first and second points mentioned above?

Thanks for the detail on the fourth point. We want our integrations to feel similar to the app that they're used to using. If "Work Items" is what they're called in Azure, it'd be great to refer to them as "Work Items" in Parabol's app too.

**Comment by nickoferrall** [comment]
Ready for a review from @Dschoordsch now

File: packages/server/postgres/migrations/1650768593875_add-AzureDevOpsDimensionFieldMap.ts
Comment (bobjac) [reply packages/server/postgres/migrations/1650768593875_add-AzureDevOpsDimensionFieldMap.ts:8]: @nickoferrall  - yes.  The additional PR for #6512 will resolve the issue you are referring to.

Comment (bobjac) [reply packages/server/postgres/migrations/1650768593875_add-AzureDevOpsDimensionFieldMap.ts:8]: I'll change it from "Issues..." to "Work Items..." on the PR for #6512.

For the issue above with the hypermedia url vs rest url, I was going to create an issue as a bug and assign to me.  I have been holding off creating 2 issues because I was going to label them as bugs, but didn't want to create a bug until it was officially in the master branch.  I plan on creating them, assigning them to me, and fixing them.

**Comment by wdfox** [comment]
> 

Bob is spot on about issue #3, though you can also resolve this directly by pulling up PGAdmin and directly fixing the spelling of the column named 'azureDevOpsFieldName' in the TaskEstimate table

**Comment by bobjac** [comment]
> Ready for a review from @Dschoordsch now

@nickoferrall - do you need anything from me to advance to the next stage, or will @Dschoordsch be notified automatically?

**Comment by nickoferrall** [comment]
> @nickoferrall - do you need anything from me to advance to the next stage, or will @Dschoordsch be notified automatically?

Nope, nothing you need to do. Georg has already been notified 👍

File: packages/server/utils/AzureDevOpsServerManager.ts
Comment (Dschoordsch) [inline packages/server/utils/AzureDevOpsServerManager.ts:261]: -1 Since you're only assigning a reference in line 260 this will modify `this.headers` which is most likely not what you want. Instead it would be easier if you would modify the next lines
```
    const res = await this.fetchWithTimeout(url, {
      ...
      headers: {
        ...this.headers,
        ['Content-Type']: 'application/json-patch+json'
      }
      ...
    })

**Review by Dschoordsch — CHANGES_REQUESTED** [review]
Some smaller issues. Everything except the migration could also be fixed in a later PR.

While testing I saw at least 2 issues:
* the issue links are wrong and lead to an API instead of the normal project page
  Issue link leads to API
* the field "Original Estimate" is also shown for Basic tasks which don't have that field
  Original Estimate field for Basic task

**Comment by bobjac** [comment]
Great.  I'll take a look this afternoon.

**Comment by bobjac** [comment]
@Dschoordsch  - the two issues you pointed out in the top comment have separate issues for them already.  
1.  For the link to the JSON object, there is an html link returned from the hypermedia REST api.  I am either going to add it as a new graphql property or replace the URL with the html link.  
2. The basic task property is going to be fixed by code I have stashed.  When testing, we need to account for the project when considering the work item type.  For example , in a Scrum project, the "bug" work item type will have different available fields than the "bug" work item type in the Agile project.  I wrote the code to capture the available fields based off the project/type combination.

File: packages/client/components/AzureDevOpsScopingSelectAllIssues.tsx
Comment (bobjac) [reply packages/client/components/AzureDevOpsScopingSelectAllIssues.tsx:107]: I removed state, teamProject, and type.  Id, title are now being used.

File: packages/client/components/AzureDevOpsFieldMenu.tsx
Comment (bobjac) [reply packages/client/components/AzureDevOpsFieldMenu.tsx:42]: Removed

File: packages/client/components/AzureDevOpsFieldMenu.tsx
Comment (bobjac) [reply packages/client/components/AzureDevOpsFieldMenu.tsx:45]: Removed

File: packages/client/types/constEnums.ts
Comment (bobjac) [reply packages/client/types/constEnums.ts:432]: Changed

File: packages/server/dataloader/azureDevOpsLoaders.ts
Comment (bobjac) [reply packages/server/dataloader/azureDevOpsLoaders.ts:419]: Work item id in the AzureDevOps API is always a number, but the official type listed in the public documentation is a string.  It is probably one of those "we reserve the right to change this to a GUID in the future".

File: packages/server/graphql/mutations/setTaskEstimate.ts
Comment (bobjac) [reply packages/server/graphql/mutations/setTaskEstimate.ts:244]: Removed

File: packages/server/graphql/mutations/updateAzureDevOpsDimensionField.ts
Comment (bobjac) [reply packages/server/graphql/mutations/updateAzureDevOpsDimensionField.ts:19]: Changed

File: packages/server/graphql/types/AzureDevOpsDimensionField.ts
Comment (bobjac) [reply packages/server/graphql/types/AzureDevOpsDimensionField.ts:24]: Changed

File: packages/server/graphql/types/AzureDevOpsDimensionField.ts
Comment (bobjac) [reply packages/server/graphql/types/AzureDevOpsDimensionField.ts:16]: changed

File: packages/server/utils/AzureDevOpsServerManager.ts
Comment (bobjac) [reply packages/server/utils/AzureDevOpsServerManager.ts:261]: I made the change.  I just want confirmation from @wdfox that this was the intent.  I will mark resolved when confirmed.

File: packages/server/postgres/queries/getAzureDevOpsDimensionFieldMaps.ts
Comment (bobjac) [reply packages/server/postgres/queries/getAzureDevOpsDimensionFieldMaps.ts:19]: Changed

File: packages/server/graphql/mutations/updateAzureDevOpsDimensionField.ts
Comment (bobjac) [reply packages/server/graphql/mutations/updateAzureDevOpsDimensionField.ts:83]: I was assuming this was the established pattern.

Comment (bobjac) [reply packages/server/graphql/mutations/updateAzureDevOpsDimensionField.ts:83]: I am happy to change if you adapt a "let it fail" pattern.

File: packages/server/graphql/mutations/updateAzureDevOpsDimensionField.ts
Comment (bobjac) [reply packages/server/graphql/mutations/updateAzureDevOpsDimensionField.ts:23]: Changed

File: packages/server/postgres/queries/upsertAzureDevOpsDimensionFieldMap.ts
Comment (bobjac) [reply packages/server/postgres/queries/upsertAzureDevOpsDimensionFieldMap.ts:11]: Changed.

**Comment by bobjac** [comment]
There was no issue to resolve in the above comment referenced at https://github.com/ParabolInc/parabol/pull/6481#discussion_r869103746, but this was copied from the previous atlassian implementation.

**Review by Dschoordsch — APPROVED** [review]
There is a small issue: when first switching to a field and voting directly (as comments is pre-selected) it fails with an error message
First estimate to comment fails
Safe to merge anyways since it's behind a feature flag.

**Comment by bobjac** [comment]
@Dschoordsch - that last issue was also related to the project/work item relationship.  I have an issue open for that.  I will create a branch and fix that since this has been merged.

**Pull request merged**
